File: services/cards-publication/src/main/java/org/opfab/cards/publication/model/RRule.java
/* Copyright (c) 2022-2024, RTE (http://www.rte-france.com)
 * See AUTHORS.txt
 * This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/.
 * SPDX-License-Identifier: MPL-2.0
 * This file is part of the OperatorFabric project.
 */

package org.opfab.cards.publication.model;

import lombok.*;
import org.springframework.validation.annotation.Validated;
import java.util.List;

@Builder
@Validated
public record RRule(FreqEnum freq, Integer count, Integer interval,
        DayEnum wkst, List<DayEnum> byweekday, List<Integer> bymonth,
        List<Integer> byhour, List<Integer> byminute, List<Integer> bysetpos,
        List<Integer> bymonthday, String tzid, Integer durationInMinutes) {
    public RRule {
        if (interval == null) {
            interval = 1;
        }
        if (wkst == null) {
            wkst = DayEnum.MO;
        }
        if (tzid == null) {
            tzid = "Europe/Paris";
        }
    }
}

File: ui/main/src/app/model/RRule.ts
/* Copyright (c) 2018-2025, RTE (http://www.rte-france.com)
 * See AUTHORS.txt
 * This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/.
 * SPDX-License-Identifier: MPL-2.0
 * This file is part of the OperatorFabric project.
 */

export class RRule {
    constructor(
        public freq?: Frequency,
        public count?: number,
        public interval?: number,
        public wkst?: Day,
        public byweekday?: Day[],
        public bymonth?: number[],
        public byhour?: number[],
        public byminute?: number[],
        public bysetpos?: number[],
        public bymonthday?: number[],
        public tzid?: string,
        public durationInMinutes?: number
    ) {}
}

export enum Frequency {
    YEARLY = 'YEARLY',
    MONTHLY = 'MONTHLY',
    WEEKLY = 'WEEKLY',
    DAILY = 'DAILY',
    HOURLY = 'HOURLY',
    MINUTELY = 'MINUTELY',
    SECONDLY = 'SECONDLY'
}

export enum Day {
    MO = 'MO',
    TU = 'TU',
    WE = 'WE',
    TH = 'TH',
    FR = 'FR',
    SA = 'SA',
    SU = 'SU'
}
